File: adapters/backend/php/src/RequireRoles.php
<?php

declare(strict_types=1);

namespace Rampart\Laravel;

use Closure;
use Illuminate\Http\JsonResponse;
use Illuminate\Http\Request;
use Psr\Http\Message\ResponseInterface;
use Psr\Http\Message\ServerRequestInterface;
use Psr\Http\Server\MiddlewareInterface;
use Psr\Http\Server\RequestHandlerInterface;

/**
 * Role-based access control middleware for Rampart.
 *
 * Must be applied after RampartMiddleware so that the 'rampart' attribute is set.
 *
 * Laravel usage:
 *   Route::middleware(['rampart', 'rampart.roles:admin,editor'])->group(function () { ... });
 *
 * PSR-15 usage:
 *   $rbac = new RequireRoles('admin', 'editor');
 *   // Stack after RampartMiddleware in your middleware pipeline.
 */
class RequireRoles implements MiddlewareInterface
{
    /** @var string[] */
    private array $requiredRoles;

    /**
     * @param string ...$roles Role names that the token must contain
     */
    public function __construct(string ...$roles)
    {
        $this->requiredRoles = $roles;
    }

    // ---------------------------------------------------------------
    // Laravel middleware interface
    // ---------------------------------------------------------------

    /**
     * Handle an incoming Laravel request.
     *
     * Laravel passes route middleware parameters as extra arguments after $next.
     * Usage: Route::middleware('rampart.roles:admin,editor')
     *
     * @param Request  $request
     * @param Closure  $next
     * @param string   ...$roles  Roles passed via route middleware definition
     * @return mixed
     */
    public function handle(Request $request, Closure $next, string ...$roles): mixed
    {
        // Use roles from route definition if provided, otherwise use constructor roles
        $required = !empty($roles) ? $roles : $this->requiredRoles;

        /** @var RampartClaims|null $claims */
        $claims = $request->attributes->get('rampart');

        if ($claims === null) {
            return new JsonResponse(
                ['detail' => 'Authentication required before role check'],
                401
            );
        }

        $missing = $claims->missingRoles(...$required);
        if (!empty($missing)) {
            return new JsonResponse(
                ['detail' => 'Missing required roles: ' . implode(', ', $missing)],
                403
            );
        }

        return $next($request);
    }

    // ---------------------------------------------------------------
    // PSR-15 MiddlewareInterface
    // ---------------------------------------------------------------

    /**
     * Process a PSR-15 server request.
     */
    public function process(ServerRequestInterface $request, RequestHandlerInterface $handler): ResponseInterface
    {
        /** @var RampartClaims|null $claims */
        $claims = $request->getAttribute('rampart');

        if ($claims === null) {
            return self::psrErrorResponse('Authentication required before role check', 401);
        }

        $missing = $claims->missingRoles(...$this->requiredRoles);
        if (!empty($missing)) {
            return self::psrErrorResponse(
                'Missing required roles: ' . implode(', ', $missing),
                403
            );
        }

        return $handler->handle($request);
    }

    /**
     * Return a PSR-7 JSON error response.
     */
    private static function psrErrorResponse(string $detail, int $status): ResponseInterface
    {
        $body = json_encode(['detail' => $detail]);
        return new \GuzzleHttp\Psr7\Response($status, ['Content-Type' => 'application/json'], $body);
    }
}

File: adapters/backend/ruby/lib/rampart/require_roles.rb
# frozen_string_literal: true

module Rampart
  # Rack middleware that enforces role-based access control.
  #
  # Must be used after {Rampart::Middleware} -- requires +env['rampart.claims']+
  # to be set. Returns 403 Forbidden if the user lacks any of the required roles.
  #
  # @example Rails route constraint
  #   # config/application.rb
  #   config.middleware.use Rampart::Middleware, issuer: "http://localhost:8080"
  #
  #   # In a specific controller or mounted Rack app:
  #   admin_app = Rampart::RequireRoles.new(app, "admin")
  class RequireRoles
    # @param app  [#call] The next Rack application in the stack.
    # @param roles [Array<String>] One or more roles the user must possess.
    def initialize(app, *roles)
      @app   = app
      @roles = roles.flatten.map(&:to_s)
    end

    def call(env)
      claims = env["rampart.claims"]

      unless claims
        return json_response(401, "unauthorized", "Authentication required.")
      end

      user_roles = Array(claims.roles)
      missing    = @roles.reject { |r| user_roles.include?(r) }

      unless missing.empty?
        return json_response(
          403, "forbidden",
          "Missing required role(s): #{missing.join(', ')}"
        )
      end

      @app.call(env)
    end

    private

    def json_response(status, error, description)
      body = JSON.generate(
        error: error,
        error_description: description,
        status: status
      )
      [
        status,
        { "content-type" => "application/json" },
        [body]
      ]
    end
  end
end
